Sources/TuistCore/Graph/GraphLoader.swift

```diff
@@ -0,0 +1,261 @@
+import Basic
+import Foundation
+import TuistSupport
+
+public protocol GraphLoading: AnyObject {
+    /// Path to the directory that contains the project.
+    /// - Parameter path: Path to the directory that contains the project.
+    func loadProject(path: AbsolutePath) throws -> (Graph, Project)
+
+    /// Loads the graph for the workspace in the given directory.
+    /// - Parameter path: Path to the directory that contains the workspace.
+    func loadWorkspace(path: AbsolutePath) throws -> (Graph, Workspace)
+
+    /// Loads the TuistConfig.
+    ///
+    /// - Parameter path: Directory from which look up and load the TuistConfig.
+    /// - Returns: Loaded TuistConfig object.
+    /// - Throws: An error if the TuistConfig.swift can't be parsed.
+    func loadTuistConfig(path: AbsolutePath) throws -> TuistConfig
+}
+
+public class GraphLoader: GraphLoading {
+    // MARK: - Attributes
+
+    fileprivate let modelLoader: GeneratorModelLoading
+
+    // MARK: - Init
+
+    public init(modelLoader: GeneratorModelLoading) {
+        self.modelLoader = modelLoader
+    }
+
+    // MARK: - GraphLoading
+
+    public func loadProject(path: AbsolutePath) throws -> (Graph, Project) {
+        let graphLoaderCache = GraphLoaderCache()
+        let graphCircularDetector = GraphCircularDetector()
+
+        let project = try loadProject(at: path, graphLoaderCache: graphLoaderCache, graphCircularDetector: graphCircularDetector)
+
+        let entryNodes: [GraphNode] = try project.targets.map { target in
+            try self.loadTarget(name: target.name, path: path, graphLoaderCache: graphLoaderCache, graphCircularDetector: graphCircularDetector)
+        }
+
+        let graph = Graph(name: project.name,
+                          entryPath: path,
+                          cache: graphLoaderCache,
+                          entryNodes: entryNodes)
+        return (graph, project)
+    }
+
+    public func loadWorkspace(path: AbsolutePath) throws -> (Graph, Workspace) {
+        let graphLoaderCache = GraphLoaderCache()
+        let graphCircularDetector = GraphCircularDetector()
+        let workspace = try modelLoader.loadWorkspace(at: path)
+
+        let projects = try workspace.projects.map { (projectPath) -> (AbsolutePath, Project) in
+            try (projectPath, self.loadProject(at: projectPath, graphLoaderCache: graphLoaderCache, graphCircularDetector: graphCircularDetector))
+        }
+
+        let entryNodes = try projects.flatMap { (project) -> [TargetNode] in
+            try project.1.targets.map { target in
```

Comment:
not a big fan of referring to tuples via implicit numbers `project.1` `project.0` for readability, can we use names?

Reply:
:+1: Pushed a commit setting a name to those variables